### ISSUE TITLE
IndexError when straightening an image using the same input image as the destination

### ISSUE THREAD

**Issue by jcohenadad** [neuropoly/spinalcordtoolbox#2527]
Hi, thanks for reporting an issue, please take some time to consider the applicable guidelines:
https://github.com/neuropoly/spinalcordtoolbox/blob/master/CONTRIBUTING.rst#reporting-a-bug-or-requesting-a-feature
(about how to suitably describe issues or requests) prior to deleting this blurb when about to submit the issue.

### Description

<Description of the issue>

### Steps to Reproduce

0. Install SCT <release or git revision and platform, you can use the
   first lines of output of sct_check_dependencies>

1. <First Step>

2. <Second Step>

3. <and so on...>

**Expected behavior:** <What you expect to happen>

**Actual behavior:** <What actually happens>

**Comment by Drulex** [comment]
I can reproduce on latest `master`. Will look into it.

**Comment by Drulex** [comment]
I don't know how productive it is for me to look at this, it seems complicated :grimacing: .

@jcohenadad can you confirm that your usage (using `-dest` with the same image as the input image) is valid? Looking at the usage help it mentions that the file passed to `-dest` should be: `Spinal cord centerline (or segmentation) of a destination image (which could be straight or curved)`. 

I'm not sure I understand how passing the same image as input satisfies this usage requirement?

**Comment by jcohenadad** [comment]
no problem i'll look into it!

**Comment by joshuacwnewton** [comment]
We know the issue is caused by the calculation trying to access an index above 240.

> ```
> IndexError: index 240 is out of bounds for axis 0 with size 240
> ```

Now, when `-dest` is passed, we see the following dimensions for the two `Centerline` objects:

![image](https://user-images.githubusercontent.com/16181459/213348459-225c2228-c495-4157-bdea-30a942b8a3b1.png)

Ah! 320!

For comparison, for the default case (no `-dest`), we see that the dimensions are both 240:

![image](https://user-images.githubusercontent.com/16181459/213353514-e59cc767-b092-47f9-befa-c795237b1d36.png)

So the question is: Why does `centerline_straight` have 320 slices when `-dest` is passed? 

----

<details>
<summary> 1. For the default case (no <code>-dest</code>), both the centerline and straightened centerline only have 240 non-null slices </summary>
-

In the straighten method, the centerline is determined by the [`_get_centerline`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/377bfb88a9845bb8348e760aea5d70bb4f4bcb42/spinalcordtoolbox/straightening.py#L620-L634) function, which then calls the `get_centerline` function, which then calls `find_and_sort_coord`:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/377bfb88a9845bb8348e760aea5d70bb4f4bcb42/spinalcordtoolbox/straightening.py#L152

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/377bfb88a9845bb8348e760aea5d70bb4f4bcb42/spinalcordtoolbox/straightening.py#L622-L623

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/377bfb88a9845bb8348e760aea5d70bb4f4bcb42/spinalcordtoolbox/centerline/core.py#L129-L130

`find_and_sort_coord` looks at only the non-null coordinates:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/377bfb88a9845bb8348e760aea5d70bb4f4bcb42/spinalcordtoolbox/centerline/core.py#L69-L70

Here, the input image has 320 SI slices, but this gets reduced down to 240 non-null slices:

![image](https://user-images.githubusercontent.com/16181459/213351887-9e63d14f-2468-4cd2-8a68-5cbe3f55197e.png)

So, by calling `_get_centerline`, we're essentially shrinking the centerline image by excluding non-null slices.

</details>


<details>
<summary> 2. In the <code>-dest</code> case, <code>t2.nii.gz</code> gets used as a straightened centerline reference, but it has 320 non-null slices instead of the expected 240 </summary>
-

Specifying `-dest` results in the following changes:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/377bfb88a9845bb8348e760aea5d70bb4f4bcb42/spinalcordtoolbox/scripts/sct_straighten_spinalcord.py#L208-L210

If we look at the `use_straight_reference` block of code, we can see the following lines:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/377bfb88a9845bb8348e760aea5d70bb4f4bcb42/spinalcordtoolbox/straightening.py#L210-L220

This call to `_get_centerline` is separate, but if we repeat the same function trace we just did with [`_get_centerline`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/377bfb88a9845bb8348e760aea5d70bb4f4bcb42/spinalcordtoolbox/straightening.py#L620-L634), we see the following:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/377bfb88a9845bb8348e760aea5d70bb4f4bcb42/spinalcordtoolbox/straightening.py#L152

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/377bfb88a9845bb8348e760aea5d70bb4f4bcb42/spinalcordtoolbox/straightening.py#L622-L623

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/377bfb88a9845bb8348e760aea5d70bb4f4bcb42/spinalcordtoolbox/centerline/core.py#L129-L130

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/377bfb88a9845bb8348e760aea5d70bb4f4bcb42/spinalcordtoolbox/centerline/core.py#L69-L70

Here, the input image has 320 SI slices, but this DOESN'T get reduced to 240 non-null slices.

![image](https://user-images.githubusercontent.com/16181459/213355420-bd4d01fd-9da3-4f41-af1f-deeb771c2d86.png)

</details>

<details>
<summary> 3. If we instead pass <code>t2_seg.nii.gz</code> to <code>-dest</code>, then there is again no dimension mismatch, and thus no error</code> </summary>

If we pass `t2_seg.nii.gz`, it is used at the centerline reference file as before, but this time, `find_and_sort_coord` properly reduces it to 240 non-null slices:

![image](https://user-images.githubusercontent.com/16181459/213357244-81ff733f-09e5-4497-8c56-3892cfc6797c.png)

And the script runs without error: https://www.diffchecker.com/4tA31No7/

</details>

----

My next question though, is: Does `-dest` _always_ have to have the exact same number of non-null slices as the detected centerline? 

`-dest` has the following usage description:

```
  -dest <file>          Spinal cord centerline (or segmentation) of a destination image (which could
                        be straight or curved). An algorithm scales the length of the input
                        centerline to match that of the destination centerline.
```

Huh. An algorithm scales the centerline! Well then, a dimension mismatch shouldn't matter?

<details>
<summary> 4. Try dilating the <code>-dest</code> image so that it's slightly bigger, but still a "centerline image"</summary>
-

Run the following:

```
sct_maths -i t2_seg.nii.gz -dilate 3 -o t2_seg_dilated.nii.gz
sct_straighten_spinalcord -i t2.nii.gz -s t2_seg.nii.gz -dest t2_seg_dilated.nii.gz 
```

Checking `find_and_sort_coord`:

![image](https://user-images.githubusercontent.com/16181459/213360275-8a933837-eca4-4648-a279-d1045dfb843f.png)

As expected, the number of non-null slices is slightly over 240. Does this cause an error? Yes!

```
IndexError: index 240 is out of bounds for axis 0 with size 240
```

</details>

<details>
<summary> 5. Try eroding the <code>-dest</code> image so that it's slightly smaller, but still a "centerline image"</summary>
-

Run the following:

```
sct_maths -i t2_seg.nii.gz -erode 3 -o t2_seg_eroded.nii.gz
sct_straighten_spinalcord -i t2.nii.gz -s t2_seg.nii.gz -dest t2_seg_eroded.nii.gz 
```

Checking `find_and_sort_coord`:

![image](https://user-images.githubusercontent.com/16181459/213361012-9ab1a200-79c1-415b-953f-b93be4977f1f.png)

As expected, the number of non-null slices is slightly under 240. Does this cause an error? Also yes!

```
IndexError: index 238 is out of bounds for axis 0 with size 237
```

</details>

So it seems like the "scaling algorithm" (whatever it may) isn't doing it's job? Right now, `-dest` must have exactly the same number of non-null slices as the detected curved centerline. That seems like an unreasonably strict requirement!

----

tl;dr I've identified the cause, but it sure doesn't seem like `-dest` is working as intended. :sweat_smile:
